feat(file storage service): add csi manila in existing k8S cluster (#8521)
ovh/docs: pages/storage_and_backup/file_storage/file_storage_service/getting_started/guide.en-gb.md

>> | `403 Forbidden`             | Project not whitelisted for Manila | Ensure you are registered to Alpha                                 |
>> | Share stuck in creating     | Invalid network ID or subnet       | Check `NETWORK_ID` and `SUBNET_ID`                                 |
>>

> > Via Manila CSI in K8s environment

Reviewer: ```suggestion
> Via Manila CSI in Kubernetes environment
```

>> **1\. Additional Requirements**
>>
>> - Helm CLI installed on your local machine.
>> - OpenStack CLI configured and ready to use.
>> - Krew (kubectl plugin manager) installed.
>> - Stern (kubectl log tailing plugin) installed via Krew.

> >> - A Kubernetes cluster deployed in a private network within a PCI region where Manila endpoints are accessible.

Reviewer: It could be interesting to indicate here that this guide works with:
* a Managed Kubernetes Service cluster
* a self-managed Kubernetes cluster

Reviewer 2: Then maybe turn it like that : 
a Managed or a self managed Kubernetes Service cluster

>>   --namespace kube-system \
>>   --version v4.11.0 \
>>   [--wait -v=5 --debug]
>> ```
>>
>> Follow the NFS CSI logs:
>>
>> ```bash
>> kubectl stern -n kube-system -l app.kubernetes.io/instance=csi-driver-nfs
>> ```
>>
>> **6\. Install the Manila CSI driver**
>>
>> Add the Helm chart repository:
>>
>> ```bash
>> helm repo add cpo https://kubernetes.github.io/cloud-provider-openstack
>> helm repo update
>> ```
>>
>> List available releases of the Manila CSI:
>>
>> ```bash
>> helm search repo -l openstack-manila-csi
>> ```
>>
>> Install the chosen Helm chart release:
>>
>> ```bash
>> helm install manila-csi cpo/openstack-manila-csi --version 2.33.1 --namespace kube-system [--values ./csi-config/helm-chart-values.yaml]
>> ```
>>
>> Follow the Manila CSI logs:
>>
>> ```bash
>> kubectl stern -n kube-system -l app=openstack-manila-csi
>> ```
>>
>> **7\. Preparing OpenStack Resources for Manila CSI**
>>
>> 1. Create a dedicated OpenStack user for Manila
>>
>> You need a separate OpenStack user to manage Manila resources. This user can be created:
>>
>> - Through the OVHcloud Control Panel (Public Cloud > Settings > Users & Roles)
>> - Or via the OVHcloud CLI
>>
>> > [!primary]
>> >

> >> > This new user must have the role share_operator or administrator.

Reviewer: This part should be confirmed with the Public Cloud team. 
The `administrator` role was required during pre-release tests but it may not the case anymore. 
IMHO, only the `share_operator` role is mandatory here.

Reviewer 2: Yes it's true, and actually, we can delete this part as when they register we give them the right role

>> >
>>
>> 2. Collect OpenStack project and user details
>>
>> Download your project’s openrc file from the OVHcloud control panel and note the following credentials:
>>
>> - os-userName
>> - os-password
>> - os-domainName
>> - os-projectDomainID
>> - os-projectName
>> - os-projectDomainID
>>
>> > [!primary]
>> >

> >> > Once you have these values, populate the csi-config/secrets.yaml file. This Kubernetes secret will allow the Manila CSI driver to manage Manila resources in your cluster.

Reviewer: It seems that the `csi-config/secret.yaml` has not been added to this PR.

>> >
>>
>> ```bash
>> curl -fsSL -o get_helm.sh https://raw.githubusercontent.com/helm/helm/main/scripts/get-helm-3
>> chmod 700 get_helm.sh
>> ./get_helm.sh
>> ```
>>

> >> Create the secret in your MKS cluster:

Reviewer: ```suggestion
>> Create the secret in your Kubernetes cluster:
```

>>
>> ```bash

> >> # Apply OpenStack credentials for Manila

Reviewer: ```suggestion
```

>> |                                   | segmentation_id = None                                   |
>> |                                   | neutron_net_id = <PRIVATE_NETWORK_NAME>                  |
>> |                                   | neutron_subnet_id = <PRIVATE_SUBNET_NAME>                |
>> |                                   | ip_version = None                                        |
>> |                                   | cidr = None                                              |
>> |                                   | network_type = None                                      |
>> |                                   | mtu = None                                               |
>> |                                   | gateway = None                                           |
>> | status                            | active                                                   |
>> | updated_at                        | None                                                     |
>> +-----------------------------------+----------------------------------------------------------+
>> ```
>>
>> **8\. Configure the Manila CSI driver**
>>

> >> Once the OpenStack share network is created, and before configuring the Manila CSI driver, you need to define the CIDR range used by your MKS nodes. This ensures that the nodes can access the Manila shares.

Reviewer: ```suggestion
>> Once the OpenStack share network is created, and before configuring the Manila CSI driver, you need to define the CIDR range used by your Kubernetes nodes. This ensures that the nodes can access the Manila shares.
```

>>
>> Edit the runtime ConfigMap:
>>

> >> Open `csi-config/manila-runtime-configmap.yaml` and update the `nfs.matchExportLocationAddress` field to match your cluster’s CIDR.

Reviewer: `csi-config/manila-runtime-configmap.yaml` file is not present in the PR.

>>
>> Example:
>>
>> ```bash
>> nfs:
>>   matchExportLocationAddress: 10.42.0.0/16
>> ```
>>
>> Apply the ConfigMap to your Kubernetes cluster:
>>
>> ```bash
>> # Optional if runtime config is already defined in Helm values
>> kubectl apply -f ./csi-config/manila-runtime-configmap.yaml
>> ```
>>
>> **9\. Creating NFS Shares via Dynamic Provisioning**
>>
>> To enable the Manila CSI driver to dynamically create Manila shares and use them as Kubernetes volumes, you must define a StorageClass in your cluster. This StorageClass specifies the shared network that will be used to create and grant access to NFS exports.
>>
>> Retrieve the shared network ID, using the OpenStack CLI to get the ID of your shared network:
>>
>> ```bash
>> openstack --os-region-name SBG5 share network list -f value -c ID
>> ```
>>
>> Configure the StorageClass:
>>

> >> - Edit the csi-config/dynamic-storageclass.yaml file.

Reviewer: File not present in the PR.

>> - Update the parameter.shareNetworkID value with the shared network ID retrieved in the previous step.
>>
>> Create the dynamic StorageClass, applying the StorageClass to your cluster:
>>
>> ```bash
>> # A StorageClass cannot be updated. If modifications are needed, delete and recreate it.
>> kubectl apply -f poc/v2/nfs/dynamic-provisioning/dynamic-storageclass.yaml
>> ```
>>
>> Once the StorageClass is created, create a `PersistentVolumeClaim` (PVC) using it. For example, claim a 15Gi volume with `ReadWriteMany` (RWX) access:
>>
>> ```bash

> >> kubectl apply -f poc/v2/nfs/dynamic-provisioning/nfs-pvc.yaml

Reviewer: File not present in the PR.

>>
>> ```bash
>> openstack --os-region SBG5 share list
>> ```
>>
>> Example output:
>>
>> ```bash
>> +--------------------------------------+------------------------------------------+------+-------------+-----------+-----------+-----------------+------+-------------------+
>> | ID                                   | Name                                     | Size | Share Proto | Status    | Is Public | Share Type Name | Host | Availability Zone |
>> +--------------------------------------+------------------------------------------+------+-------------+-----------+-----------+-----------------+------+-------------------+
>> | 9484d5f3-7bf7-486b-b88e-40bbedeet9f3 | pvc-78135a68-c6f4-48fe-8644-454b387a3ad4 |   15 | NFS         | available | False     | generic_0       |      | nova              |
>> +--------------------------------------+------------------------------------------+------+-------------+-----------+-----------+-----------------+------+-------------------+
>> ```
>>
>> Deploy a Kubernetes Deployment for a web server with 2 replicas. The pods will mount and share the previously created `RWX` volume:
>>
>> ```bash

> >> kubectl apply -f poc/v2/nfs/dynamic-provisioning/nfs-deployment.yaml

Reviewer: File not present in the PR.

>>
>> As shown earlier, a Kubernetes StorageClass can dynamically create Manila shares exposed via NFS. Alternatively, you can use a pre-provisioned Manila share and mount it directly in a Pod.
>>
>> Create a new share with the OpenStack CLI:
>>
>> ```bash
>> openstack --os-region SBG5 share create \
>>   --share-type <SHARE_TYPE> \
>>   --share-network <SHARE_NETWORK_ID> \
>>   --name <NFS_SHARE_NAME> \
>>   SHARE_PROTOCOL SHARE_SIZE
>> ```
>>
>> Where:
>>
>> - `SHARE_TYPE` is `generic_0` by default. You can check existing share types with:
>>
>> ```bash
>> openstack --os-region SBG5 share type list
>> ```
>>
>> - `SHARE_NETWORK_ID` is the identifier of your shared network.
>> - `NFS_SHARE_NAME` is the name of your NFS share.
>> - `SHARE_PROTOCOL` is the protocol to use (currently, only NFS is supported).
>> - `SHARE_SIZE` is the size to allocate to your NFS volume (in GiB).
>>
>> Create a share access rule:
>>
>> ```bash
>> openstack --os-region SBG5 share access create <NFS_SHARE_NAME> ip <SUBNET_CIDR>
>> ```
>>
>> Where:
>>
>> - `SHARE_ACCESS_NAME` is the name of the share.
>> - `SUBNET_CIDR` is the CIDR used when configuring the Manila CSI runtime.
>>

> >> Retrieve the NFS share ID and the share access ID, then update the `volumeAttributes.shareID` and `volumeAttributes.shareAccessID` parameters in the `poc/v2/nfs/static-provisioning/static-provisioning.yaml` file.

Reviewer: File not present in the PR.

>> - Or via the OVHcloud CLI
>>
>> 2. Collect OpenStack project and user details
>>
>> Download your project’s openrc file from the OVHcloud control panel and note the following credentials:
>>
>> - os-userName
>> - os-password
>> - os-domainName
>> - os-projectDomainID
>> - os-projectName
>>
>> Once you have these values, create a file named secrets.yaml with the following content. This Kubernetes Secret allows the Manila CSI driver to authenticate against OpenStack and manage Manila resources in your cluster.
>>
>> ```yaml
>> apiVersion: v1
>> kind: Secret
>> metadata:
>>   name: csi-manila-secrets
>>   namespace: default
>> stringData:
>>   # Mandatory
>>   os-authURL: "https://auth.cloud.ovh.net/v3"
>>   os-region: "SBG5"
>>
>>   # Openstack authentication
>>   os-userName: "<os-userName>"
>>   os-password: "<os-password>"
>>   os-domainName: "default"
>>   os-projectName: "os-projectName"
>>   os-projectDomainID: "default"
>> 
>>   # Authentication using trustee credentials
>>   # os-trustID: "some-trust-id"
>>   # os-trusteeID: "some-trustee-id"
>>   # os-trusteePassword: "some-trustee-password"
>> ```
>>
>> Then, apply it to your Kubernetes cluster:
>>
>> ```bash

> >> kubectl apply -f ./secrets.yaml

Reviewer: ```suggestion
>> kubectl apply -f secrets.yaml
```

>>
>> Edit the runtime ConfigMap:
>>
>> Create a file named `manila-runtime-configmap.yaml` and update the `nfs.matchExportLocationAddress` field to match your cluster’s CIDR:
>>
>> ```yaml
>> apiVersion: v1
>> kind: ConfigMap
>> metadata:
>>   name: manila-csi-runtimeconf-cm
>>   namespace: default
>>   annotations:
>>     meta.helm.sh/release-name: manila-csi
>>     meta.helm.sh/release-namespace: default
>>   labels:
>>     app.kubernetes.io/managed-by: Helm
>> data:
>>   runtimeconfig.json: |
>>     {
>>       "nfs": {
>>         # When mounting an NFS share, select an export location with
>>         # this IP address. No match between this address and
>>         # at least a single export location for this share will
>>         # result in an error.
>>         # Expects a CIDR-formatted address. If prefix is not provided,
>>         # /32 or /128 prefix is assumed for IPv4 and IPv6 respectively.
>>         "matchExportLocationAddress": "<SUBNET_CIDR>"
>>       }
>>     }
>> ```
>>
>> Example:
>>
>> ```bash

> >> nfs:

Reviewer: ```suggestion
>> "nfs":
```

> >>   matchExportLocationAddress: 10.42.0.0/16

Reviewer: ```suggestion
>>   "matchExportLocationAddress": "10.42.0.0/16"
```

>> ```
>>
>> Apply the ConfigMap to your Kubernetes cluster:
>>
>> ```bash
>> # Optional if runtime config is already defined in Helm values

> >> kubectl apply -f ./manila-runtime-configmap.yaml

Reviewer: ```suggestion
>> kubectl apply -f manila-runtime-configmap.yaml
```

>> ```
>>
>> Configure the StorageClass:
>>
>> - Create a file named `dynamic-storageclass.yaml` with the following content:
>>
>> ```yaml
>> apiVersion: storage.k8s.io/v1
>> kind: StorageClass
>> metadata:
>>   name: csi-manila-nfs
>> provisioner: nfs.manila.csi.openstack.org
>> allowVolumeExpansion: true
>> parameters:
>>   # Manila share type
>>   # default value: default
>>   # openstack share type list to find proper value
>>   type: generic_0
>>   # /!\ MANDATORY /!\
>>   # openstack share network list
>>   shareNetworkID: "<OS_SHARE_NETWORK_ID>"
>>   nfs-shareClient: "<SUBNET_CIDR>"
>>
>>   csi.storage.k8s.io/provisioner-secret-name: csi-manila-secrets
>>   csi.storage.k8s.io/provisioner-secret-namespace: default
>>   csi.storage.k8s.io/controller-expand-secret-name: csi-manila-secrets
>>   csi.storage.k8s.io/controller-expand-secret-namespace: default
>>   csi.storage.k8s.io/node-stage-secret-name: csi-manila-secrets
>>   csi.storage.k8s.io/node-stage-secret-namespace: default
>>   csi.storage.k8s.io/node-publish-secret-name: csi-manila-secrets
>>   csi.storage.k8s.io/node-publish-secret-namespace: default
>> ```
>>

> >> - Update the parameter.shareNetworkID value with the shared network ID retrieved in the previous step.

Reviewer: ```suggestion
>> * update the `parameter.shareNetworkID` value with the shared network identifier.
>> * update the `parameter.nfs-shareClient` value with the subnet CIDR defined during cluster creation.
```

>> - os-domainName
>> - os-projectDomainID
>> - os-projectName
>>
>> Once you have these values, create a file named secrets.yaml with the following content. This Kubernetes Secret allows the Manila CSI driver to authenticate against OpenStack and manage Manila resources in your cluster.
>>
>> ```yaml
>> apiVersion: v1
>> kind: Secret
>> metadata:
>>   name: csi-manila-secrets
>>   namespace: default
>> stringData:
>>   # Mandatory
>>   os-authURL: "https://auth.cloud.ovh.net/v3"
>>   os-region: "SBG5"
>>
>>   # Openstack authentication
>>   os-userName: "<os-userName>"
>>   os-password: "<os-password>"
>>   os-domainName: "default"
>>   os-projectName: "os-projectName"
>>   os-projectDomainID: "default"
>> 

> >>   # Authentication using trustee credentials

Reviewer: Maybe we could remove the "trustID" configuration part?
